src/hotspot/share/code/compressedStream.cpp

```diff
@@ -112,3 +112,89 @@ void CompressedWriteStream::write_long(jlong value) {
   write_signed_int(low(value));
   write_signed_int(high(value));
 }
+
+
+bool CompressedSparseDataReadStream::read_zero() {
+  if (_buffer[_position] & (1 << (7 - _bit_pos))) {
+    return 0; // not a zero data
+  }
+  if (++_bit_pos == 8) {
+    _position++;
+    _bit_pos = 0;
+  }
+  return 1;
+}
+
+uint8_t CompressedSparseDataReadStream::read_byte_impl() {
+  uint8_t b1 = _buffer[_position] << _bit_pos;
+  uint8_t b2 = _buffer[++_position] >> (8 - _bit_pos);
+  return b1 | b2;
+}
+
+jint CompressedSparseDataReadStream::read_int() {
+  if (read_zero()) {
+    return 0;
+  }
+  // integer value encoded as a sequence of 1 to 5 bytes
+  // - the most frequent case (0 < x < 64) is encoded in one byte
+  // - the payload of the first byte is 6 bits, the payload of the following bytes is 7 bits
+  // - the most significant bit in the first byte is occupied by a zero flag
+  // - each byte has a bit indicating whether it is the last byte in the sequence
+  uint8_t b = read_byte_impl();
+  juint result = b & 0x3f;
+  for (int i = 0; (i == 0) ? (b & 0x40) : (b & 0x80); i++) {
+    b = read_byte_impl();
+    result |= ((b & 0x7f) << (6 + 7 * i));
+  }
+  return (jint)result;
+}
+
+int CompressedSparseDataWriteStream::position() {
+  if (_bit_pos == 0) {
+    return _position;
+  }
+  // flush current data and start a new byte
+  write(_curr_byte << (8 - _bit_pos));
+  _curr_byte = 0;
+  _bit_pos = 0;
+  return _position;
+}
+
+void CompressedSparseDataWriteStream::write_zero() {
+  _curr_byte <<= 1; // zero bit represents a zero word
+  if (++_bit_pos == 8) {
+    write(_curr_byte);
+    _curr_byte = 0;
+    _bit_pos = 0;
+  }
+}
+
+void CompressedSparseDataWriteStream::write_byte_impl(uint8_t b) {
+  write((_curr_byte << (8 - _bit_pos)) | (b >> _bit_pos));
+  _curr_byte = (0xff >> (8 - _bit_pos)) & b;
+}
+
+void CompressedSparseDataWriteStream::write_int(juint val) {
+  if (val == 0) {
+    write_zero();
+    return;
+  }
+  int bit0 = 0x80; // first byte upper bit is set to indicate a value is not zero
+  juint next = val >> 6;
+  int bit1 = (next != 0) ? 0x40 : 0; // bit indicating a last byte
+  write_byte_impl(bit0 | bit1 | (val & 0x3f));
+  while (next != 0) {
+    bit1 = (next >> 7) ? 0x80 : 0;
+    write_byte_impl(bit1 | (next & 0x7f));
+    next >>= 7;
+  }
```

Comment:
Could you please document the encoding/decoding schema in comments to the class? Could you also please include examples?
Why is it chosen to split 32 bits into 6, 7, 7, 7 and 5?